### PR TITLE
Tests for collect and squashoutput

### PR DESCRIPTION
**Pull request by johnomotani**
Adds tests for `collect()` and `squashoutput()` to the test suite. The tests cover core, SOL, single-null, connected double-null and disconnected double-null topologies (the last two aren't really different because `collect()` doesn't need the separatrix positions `ixseps1` and `ixseps2`, but is included for completeness - limiter topology is missing, but is not really different from core or SOL for the same reason). The `xguards`, `yguards`, `tind`, `xind`, `yind` and `zind` arguments are all tested, as are the `compression` and `complevel` arguments to `squashoutput` (which verify that we get identical data when using compression). Edit: now added tests for different `MXG` and `MYG` values for one case.

If anyone can see anything significant missing, please add it!

### CODE REVIEW

**Comment by johnomotani** [comment]
One downside is that the tests are a bit slow (CI takes ~15mins), but I think that's inevitable when reading/writing to/from disk - I think it's more important to test as many arguments to `collect` and `squashoutput` as possible to catch edge cases.

**Comment by loeiten** [comment]
What an awesome piece of work @johnomotani 🥳
I can have a look at it once I get some time at my hands

File: boutdata/tests/make_test_data.py [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+def create_dump_file(*, i, tmpdir, rng, grid_info, boundaries, fieldperp_global_yind):
+    """
+    Create a netCDF file mocking up a BOUT++ output file, and also return the data
+    without guard cells
+
+    Parameters
+    ----------
+    i : int
+        Number of the output file
+    tmpdir : pathlib.Path
```
Comment (loeiten) [inline boutdata/tests/make_test_data.py:88]: Looks like this line is duplicated
```suggestion
```

File: boutdata/tests/make_test_data.py [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+
+    # Scalar
+    def createScalar(name, value):
+        var = outputfile.createVariable(name, type(value))
+
+        var[...] = value
+
+        result[name] = value
+
+    createScalar("BOUT_VERSION", 4.31)
```
Comment (loeiten) [inline boutdata/tests/make_test_data.py:240]: I guess the version number needs to be updated manually if needed?

File: boutdata/tests/make_test_data.py [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+        createScalar(key, value)
+    nxpe = grid_info["NXPE"]
+    createScalar("PE_XIND", i % nxpe)
+    createScalar("PE_YIND", i // nxpe)
+    createScalar("MYPE", i)
+
+    return result
+
+
+def concatenate_data(data_list, *, nxpe, fieldperp_yproc_ind):
```
Comment (loeiten) [inline boutdata/tests/make_test_data.py:251]: Would a short docstring be helpful here?
Is it for example correct that `data_list` is a list where the index refers to the processor number and the value is a dict on the `expected_attrbutes` form?

File: boutdata/tests/make_test_data.py [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+                    fieldperp_yproc_ind * nxpe, (fieldperp_yproc_ind + 1) * nxpe
+                )
+            ],
+            axis=0,
+        )
+
+    return result
+
+
+def apply_slices(expected, tslice, xslice, yslice, zslice):
```
Comment (loeiten) [inline boutdata/tests/make_test_data.py:312]: If you'd time to write a short docstring here as well, that would be great. Can for example `tslice` be a `slice` object?

File: boutdata/tests/make_test_data.py [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+        )
+
+    for varname in field3d_list + field2d_list:
+        expected[varname] = np.concatenate(
+            [
+                expected[varname][:, : ny_inner + myg],
+                expected[varname][:, ny_inner + 3 * myg :],
+            ],
+            axis=1,
+        )
```
Comment (loeiten) [inline boutdata/tests/make_test_data.py:390]: I haven't checked the slicings, but I'm assuming they are good

File: boutdata/tests/test_collect.py [hunk 6]
```diff
@@ -0,0 +1,2773 @@
...
+def check_collected_data(
+    expected,
+    *,
+    fieldperp_global_yind,
+    doublenull,
+    path,
+    squash,
+    collect_kwargs,
+    squash_kwargs={},
+):
```
Comment (loeiten) [inline boutdata/tests/test_collect.py:41]: A docstring here as well would be great 😊

File: boutdata/tests/test_collect.py [hunk 7]
```diff
@@ -0,0 +1,2773 @@
...
+        remove_yboundaries(expected, expected["MYG"], expected["ny_inner"], doublenull)
+
+    collect_kwargs = collect_kwargs.copy()
+    if squash:
+        squashoutput(path, outputname="boutdata.nc", **collect_kwargs)
+        collect_kwargs["prefix"] = "boutdata"
+        # Delete dump files to be sure we do not read from them
+        dump_names = glob(str(path.joinpath("BOUT.dmp.*.nc")))
+        for x in dump_names:
+            Path(x).unlink()
```
Comment (loeiten) [inline boutdata/tests/test_collect.py:59]: I guess this is ok when running tests in parallel as the `tmp_path` fixture is being used?

File: boutdata/tests/test_collect.py [hunk 8]
```diff
@@ -0,0 +1,2773 @@
...
+        grid_info["ny"] = grid_info["NYPE"] * grid_info["MYSUB"]
+        grid_info["nz"] = grid_info["NZPE"] * grid_info["MZSUB"]
+        grid_info["MZ"] = grid_info["nz"]
+        grid_info["ixseps1"] = 7
+        grid_info["ixseps2"] = 7
+        grid_info["jyseps1_1"] = -1
+        grid_info["jyseps2_1"] = grid_info["ny"] // 2 - 1
+        grid_info["ny_inner"] = grid_info["ny"] // 2
+        grid_info["jyseps1_2"] = grid_info["ny"] // 2 - 1
+        grid_info["jyseps2_2"] = grid_info["ny"] // 2
```
Comment (loeiten) [inline boutdata/tests/test_collect.py:127]: It looks like `grid_info` is only slightly modified from time to time. Would it make sense to make `grid_info` a class member and let each test make the needed modifications on a copy?

File: boutdata/tests/test_collect.py [hunk 9]
```diff
@@ -0,0 +1,2773 @@
...
+        elif "_t" in varname or varname == "t_array":
+            assert actual.attributes["bout_type"] == "scalar_t"
+        else:
+            assert actual.attributes["bout_type"] == "scalar"
+
+
+class TestCollect:
+    @pytest.mark.parametrize("squash", [False, True])
+    @pytest.mark.parametrize("collect_kwargs", collect_kwargs_list)
+    def test_core_min_files(self, tmp_path, squash, collect_kwargs):
```
Comment (loeiten) [inline boutdata/tests/test_collect.py:105]: Although the function names are quite descriptive, I've found simple one-liners describing the tests very helpful in other projects

File: boutdata/tests/test_collect.py [hunk 10]
```diff
@@ -0,0 +1,2773 @@
...
+
+        fieldperp_global_yind = 3
+        fieldperp_yproc_ind = 0
+
+        rng = np.random.default_rng(101)
+
+        # core
+        # core includes "ylower" and "yupper" even though there is no actual y-boundary
+        # because collect/squashoutput collect these points
+        dump_params = [
```
Comment (loeiten) [inline boutdata/tests/test_collect.py:202]: Is `dump_params` also often duplicated in the tests? Would it make sense to have a class member which is only slightly changed from time to time?

**Review by loeiten — COMMENTED** [review]
Looks good as far as I can see.
Only minor suggestions

File: boutdata/tests/make_test_data.py (hunk 2 above)
Comment (johnomotani) [reply boutdata/tests/make_test_data.py:240]: Even better might be to parameterise it. At the moment though, I think the only reason it needs to be there is that there's a branch to support BOUT++ versions 3 and earlier (which removes the last point in the z-direction), so we need a version number >=4. If we add more version-specific behaviour in future it might be good to actually test that.

File: boutdata/tests/test_collect.py (hunk 7 above)
Comment (johnomotani) [reply boutdata/tests/test_collect.py:59]: Yes, quoting the docs https://docs.pytest.org/en/6.2.x/tmpdir.html: "the tmp_path fixture which will provide a temporary directory unique to the test invocation, created in the base temporary directory."

File: boutdata/tests/test_collect.py (hunk 8 above)
Comment (johnomotani) [reply boutdata/tests/test_collect.py:127]: Good point. There was a small amount of logic (depending on number of X-points) in creating `grid_info`, so I've added a utility method to construct the grid_info dicts. Reduces the number of lines quite a bit :+1:

File: boutdata/tests/test_collect.py (hunk 10 above)
Comment (johnomotani) [reply boutdata/tests/test_collect.py:202]: I think it makes sense to keep `dump_params` as is. Those parameters are setting out the processor layout manually in a reasonably compact way. I don't want to replicate logic from `collect()` when testing `collect()`, so I think giving them explicitly in each test is the best.

**Comment by johnomotani** [comment]
Thanks @loeiten! I think I've addressed all your comments now.

File: boutdata/tests/make_test_data.py (hunk 2 above)
Comment (loeiten) [reply boutdata/tests/make_test_data.py:239]: That makes sense 🙂

File: boutdata/tests/test_collect.py (hunk 10 above)
Comment (loeiten) [reply boutdata/tests/test_collect.py:266]: Yes, after looking again, it looks like they are varying quite a bit

**Review by loeiten — APPROVED** [review]
LGTM. Great work 😃
